### DISCUSSION TITLE
YLearn for Timeseries data

### DISCUSSION THREAD

**Discussion by CJames1261** [DataCanvasIO/YLearn#55] [General]
I have a time series dataset and am wondering how I could use YLearn in with my data? For example lets say I have 4 columns: rate1,rate2,rate3,rate4 (which are all floats). and lets say I wanted to find the causal estimator score for how each rate affects my 'rate1' column (my target feature).

After running the why.fit() function like so 'why.fit(train_data,outcome)', I get my treatment float columns, but am unsure how I can see visualizations or find insights into why/how these treatment variables affect my target feature. how can I find the causal effect of my treatment variables on my outcome variable, and create visualizations with meaningful insights? Thanks!

**Comment by bochen-lyu** [comment]
Hi,

Thanks for the question. Currently YLearn is not specially designed for time series data, thus it might not be appropriate to directly apply YLearn to time series data.

However, it is still possible to apply YLearn to the example data suggested in your comment.  Specifically, when calling ``why.fit()``, it is better to specify the value of ``treatment``, e.g., ``why.fit(train_data, outcome, treatment=['rate2', 'rate3'])``. Then you can find the causal effect of treatment ``rate2`` and ``rate3`` on the outcome ``rate1`` by calling ``why.causal_effect()``.

**Reply by CJames1261** [reply]
Thank you for that response! 

Th at is helpful to know. I haven't been able to find a time series causal analysis library. If you have any suggestions or examples then I'm all ears. 

one more question if you will: How do I interpret the causality score results? For example looking at the california housing example I get these results:

"""
why = Why()
why.fit(data, outcome, treatment=['AveBedrms', 'AveRooms'])

print(why.causal_effect())
"""

Outcome: 
               mean       min        max       std
AveBedrms -0.006393 -1.264421  11.482253  0.132226
AveRooms  -0.068143 -1.728602   0.748265  0.054556

How do I interpret these result on my outcome variable? (For example is this saying that for every unit increase in aveBedrms there is a -.006 decrease in outcome variable? Thanks!

**Reply by CJames1261** [reply]
Also Kudos for making this library, its very useful!

**Reply by bochen-lyu** [reply]
> If you have any suggestions or examples then I'm all ears.

It seems that there indeed lacks such python lib. Currently we only support the causal discovery methods for time series data, perhaps we might start working on including more supports of time series data after a comprehensive investigation.

> How do I interpret these result on my outcome variable? (For example is this saying that for every unit increase in aveBedrms there is a -.006 decrease in outcome variable? Thanks!

Yes, this interpretation is basically correct. You can also check our [doc](https://ylearn.readthedocs.io/en/latest/sub/est.html) for a helpful definition of the average treatment effect (ATE) to get a sense about the interpretation of causal effects in your case. 

> Also Kudos for making this library, its very useful!

Thanks for saying this 😄

**Reply by CJames1261** [reply]
Hi!

I apologize for asking so many questions, I deleted my last two posts because I was able to figure them out after looking at your pyscripts. I did have one suggestion/question however:

Is there anyway to know the columns that are associated to their coefficient values when using the code: `double_ml_model.yx_model.coef_`?
I have a client who is interested in viewing the coefficients with their associated column names. I also think this would be a helpful feature especially when there are multiple treatment columns being fitted in the model. 

I also have been able to create a a tutorial .py script for my colleagues who are interested in this library. If you would like to look at the script and potentially put it in your example notebooks, then please let me know. 

Thank you again for your time!

**Reply by bochen-lyu** [reply]
Hi,

Sorry for the late reply and thanks for your question. 

> Is there anyway to know the columns that are associated to their coefficient values when using the code: ``double_ml_model.yx_model.coef_``?

In fact, ``yx_model.coef_`` should not be directly used when viewing the causal effect and the corresponding ``coef_`` does not directly relate to the features (also column names) in the given data. This is because the corresponding ``coef_`` are obtained in a "tricky" way from the given data and is only an intermediate step to estimate the final causal effect. 

To view the causal effect, you can directly use the function ``DoubleML.estimate()``. Please see this [doc](https://ylearn.readthedocs.io/en/latest/sub/est_model/dml.html) if you are interested in how the ``coef_`` is related to the data and  causal effect, where we also specify how to use the ``DoubleML.estimate()`` function when there are multiple treatment columns:
> ``treat (float or numpy.ndarray, optional, default=None)`` – In the case of single discrete treatment, treat should be an int or str of one of all possible treatment values which indicates the value of the intended treatment; in the case of multiple discrete treatment, treat should be a list or an ndarray where treat[i] indicates the value of the i-th intended treatment, for example, when there are multiple discrete treatments, array([‘run’, ‘read’]) means that the treat value of the first treatment is taken as ‘run’ and that of the second treatment is taken as ‘read’; in the case of continuous treatment, treat should be a float or an ndarray.

Please also see the [example notebook](https://github.com/DataCanvasIO/YLearn/blob/main/example_usages/test_dml.ipynb) for several cases of applying ``DoubleML``.

----
> I also have been able to create a a tutorial .py script for my colleagues who are interested in this library. If you would like to look at the script and potentially put it in your example notebooks, then please let me know.

It would be great to have this tutorial! Could you please let us look at this script first?

Best regards

**Reply by CJames1261** [reply]
Hi, 

I am a little confused on the _cross_fit function. I'm just a little confused on exactly what is going on, maybe you could help clarify.

I am using this code: 'double_ml_model.fit(data,outcome,treatment=['MedInc', 'HouseAge'],covariate=['AveRooms', 'AveBedrms', 'Population', 'AveOccup', 'Latitude', 'Longitude'])' 

I understand there is an x_model being created, but I'm a little confused on what exactly the x_model is. Are individual models for each treatment being created with 'wv' columns  as the independent variables? And for each fold is each treatment being predicted.

I'm a little confused because when I print the 'target_converted = target' varible in the first else statement I get these (almost looks like concatenated) values:
[[ 8.3252 41.    ]
 [ 8.3014 21.    ]
 [ 7.2574 52.    ]
 ...
 [ 1.7    17.    ]
 [ 1.8672 18.    ]
 [ 2.3886 16.    ]]
 
 I know these values are values form the two treatments variables, so is the model trying to predict both of these values on each fold it is training and testing? does it see these as one value? I am familiar with machine learning, but I'm  used to seeing it in the traditional example way that sklearn has them set up. I also am not an expert programmer so I could also be confusing myself. Thanks!

**Comment by Elon991613** [comment]
Hello, I have a question for you. 

In this case, I found that the main focus is on studying the causal effect of each attribute in a certain feature variable (such as card type) on the transaction amount. Can Ylearn explain the causal effect between each feature variable on the transaction amount? Instead of just analyzing one of the feature variables, a more macro approach is used to perform a causal inference on all variables to obtain the causal effect of each feature variable on the result variable. 

Thank you for your answer.
